Convert Editor to functional, splits off CodeMirror code into its own file

khanniie opened the pull request:
Part of larger CodeMirror v5 -> v6 conversion. This converts the Editor component from class-based to functional style.

Changes:

I have verified that this pull request:

* [x] has no linting errors (`npm run lint`)
* [x] has no test errors (`npm run test`)
* [x] is from a uniquely-named feature branch and is up to date with the  `develop` branch.
* [ ] is descriptively named and links to an issue number, i.e. `Fixes #123`

Vishawdeep-Singh committed:
fix Exiting Multi-Line Cursor Doesnt Return to the Original Cursor

raclim committed:
remove extra console log

khanniie committed:
pull out codemirror test

khanniie committed:
Convert editor file to functional

raclim commented:
Thanks so much for your work on this! After doing some manual testing, I think this mostly felt good besides an issue with unsaved changes—I attached some videos below to help visualize it!

 In the current setup, if I have unsaved changes, the changes would remain. 

https://github.com/user-attachments/assets/104430da-9678-43c1-9618-e353b68ea983


Here, I think I'm seeing that the unsaved changes disappear on Autosave. 
 

https://github.com/user-attachments/assets/e8bd4958-cee8-4bd7-9b0c-28fe2a31a8c4



Besides this, I feel like everything else seems to work well for me!

khanniie committed:
change CM fifle to a useCodeMirror custom hook

khanniie committed:
fix codemirror file debounce bug

khanniie committed:
fix codemirrorr bug with debouncer again

khanniie committed:
only useeffect on file id change

khanniie committed:
switches from using usePrevious to an existing hook

khanniie committed:
moves document management to the cm file

khanniie committed:
fix test error

khanniie committed:
fix mobile bug

khanniie committed:
adds comments, removes a logic block that wasn't doing anything

I'm fairly certain that we can remove ```
      if (!prevProps?.unsavedChanges) {
        setTimeout(() => setUnsavedChanges(false), 400);
      }
```

I looked at the git blame and it looks like the intention was to stop setting unsavedchanges to false when the files switched, but i think the solve that was implemented 9 years ago did something like "if c == false, c = false" and we should be able to safely remove it instead.

reference commit: https://github.com/processing/p5.js-web-editor/commit/77e2f5bfff4774d52adc0ec2c8dd06fd22bd7507

khanniie committed:
add TODO

raclim reviewed (approved):
Thanks so much for your work on this!!! This overall looks good to me! :) 

I feel like any further issues that come up from this will probably be on the minor end, and can be addressed down the line!
